File: src/main/java/com/myudog/myulib/internal/event/EventDispatcherImpl.java
package com.myudog.myulib.internal.event;

import com.myudog.myulib.api.event.Event;
import com.myudog.myulib.api.event.EventPriority;
import com.myudog.myulib.api.event.ProcessResult;
import com.myudog.myulib.api.event.listener.EventListener;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class EventDispatcherImpl {
    private static final class PrioritizedListener<T extends Event> implements Comparable<PrioritizedListener<T>> {
        private final int priority;
        private final EventListener<T> listener;

        private PrioritizedListener(int priority, EventListener<T> listener) {
            this.priority = priority;
            this.listener = listener;
        }

        @Override
        public int compareTo(PrioritizedListener<T> other) {
            return Integer.compare(this.priority, other.priority);
        }
    }

    private final Map<Class<? extends Event>, List<PrioritizedListener<? extends Event>>> listeners = new HashMap<>();

    public <T extends Event> EventListener<T> subscribe(Class<T> eventType, EventListener<T> listener) {
        return subscribe(eventType, EventPriority.NORMAL, listener);
    }

    public <T extends Event> EventListener<T> subscribe(Class<T> eventType, int priority, EventListener<T> listener) {
        List<PrioritizedListener<? extends Event>> list = listeners.computeIfAbsent(eventType, key -> new ArrayList<>());
        list.add(new PrioritizedListener<>(priority, listener));
        list.sort(Comparator.comparingInt(entry -> entry.priority));
        return listener;
    }

    public <T extends Event> void unsubscribe(Class<T> eventType, EventListener<T> listener) {
        List<PrioritizedListener<? extends Event>> list = listeners.get(eventType);
        if (list == null) {
            return;
        }
        list.removeIf(entry -> entry.listener.equals(listener));
        if (list.isEmpty()) {
            listeners.remove(eventType);
        }
    }

    public ProcessResult dispatch(Event event) {
        List<PrioritizedListener<? extends Event>> list = listeners.get(event.getClass());
        if (list == null || list.isEmpty()) {
            return ProcessResult.PASS;
        }

        List<PrioritizedListener<? extends Event>> snapshot = new ArrayList<>(list);
        for (PrioritizedListener<? extends Event> wrapper : snapshot) {
            @SuppressWarnings("unchecked")
            PrioritizedListener<Event> typedWrapper = (PrioritizedListener<Event>) wrapper;
            ProcessResult result = typedWrapper.listener.handle(event);
            if (result != ProcessResult.PASS) {
                return result;
            }
        }
        return ProcessResult.PASS;
    }

    public void clear() {
        listeners.clear();
    }
}

File: src/main/kotlin/com/myudog/myulib/internal/event/EventDispatcherImpl.kt
package com.myudog.myulib.internal.event

import com.myudog.myulib.api.event.Event
import com.myudog.myulib.api.event.EventPriority
import com.myudog.myulib.api.event.ProcessResult
import com.myudog.myulib.api.event.listener.EventListener
import kotlin.reflect.KClass

private class PrioritizedListener<T : Event>(
    val priority: UInt,
    val listener: EventListener<T>
) : Comparable<PrioritizedListener<T>> {
    override fun compareTo(other: PrioritizedListener<T>): Int {
        return this.priority.compareTo(other.priority)
    }
}

/**
 * 改為 open class，允許被實例化。不再使用 object 全域綁死。
 */
open class EventDispatcherImpl {
    // 每個實例都有自己獨立的 listeners Map
    private val listeners = mutableMapOf<KClass<out Event>, MutableList<PrioritizedListener<out Event>>>()

    inline fun <reified T : Event> subscribe(
        priority: UInt = EventPriority.NORMAL,
        listener: EventListener<T>
    ): EventListener<T> {
        return subscribe(T::class, priority, listener)
    }

    fun <T : Event> subscribe(
        eventType: KClass<T>,
        priority: UInt,
        listener: EventListener<T>
    ): EventListener<T> {
        val list = listeners.getOrPut(eventType) { mutableListOf() }
        list.add(PrioritizedListener(priority, listener))

        @Suppress("UNCHECKED_CAST")
        (list as MutableList<PrioritizedListener<T>>).sort()

        return listener
    }

    inline fun <reified T : Event> unsubscribe(listener: EventListener<T>) {
        unsubscribe(T::class, listener)
    }

    fun <T : Event> unsubscribe(eventType: KClass<T>, listener: EventListener<T>) {
        listeners[eventType]?.removeIf { it.listener == listener }
    }

    fun dispatch(event: Event): ProcessResult {
        val eventType = event::class
        val list = listeners[eventType] ?: return ProcessResult.PASS

        for (wrapper in list) {
            @Suppress("UNCHECKED_CAST")
            val typedWrapper = wrapper as PrioritizedListener<Event>
            val result = typedWrapper.listener.handle(event)
            if (result != ProcessResult.PASS) {
                return result
            }
        }
        return ProcessResult.PASS
    }
}